### PR TITLE
fix #9503 plus another proposal on comparison and iteration syntax

### LINKED ISSUE #9503: produce only one Expr type from A <: B expressions
As discussed in #9474, the parsing of `A <: B` is currently somewhat inconsistent, and this makes metaprogamming a bit harder than necessary:
```
julia> dump(:(A <: B))
Expr 
  head: Symbol comparison
  args: Array(Any,(3,))
    1: Symbol A
    2: Symbol <:
    3: Symbol B
  typ: Any

julia> dump(:(type A <: B; end))
Expr 
  head: Symbol type
  args: Array(Any,(3,))
    1: Bool true
    2: Expr 
      head: Symbol <:
      args: Array(Any,(2,))
        1: Symbol A
        2: Symbol B
      typ: Any
    3: Expr 
      head: Symbol block
      args: Array(Any,(0,))
      typ: Any
  typ: Any
```
Perhaps `A <: B` should be parsed as `Expr(:(<:), :A, :B)` in all cases.  cc: @yuyichao

### PR DESCRIPTION
**Pull request by JeffBezanson**
The first commit is a straightforward fix for #9503, always parsing `a<:b` as an expression with head `<:`.

This change inspired the second commit, which does two related things. First, 2-argument comparisons are parsed as calls. Previously, we parsed `a<b` as `(comparison a < b)`, which I think is a quasi-bug since this can be handled as a simple function call and doesn't need chained comparison treatment.

That led to looking at iteration syntax, which depends on the parsing of `a in b`. One problem here is that `=` and `in` have different precedence, so depending on which you used you could get different syntax errors. For example `for i = a&&b` parses but `for i in a&&b` is a syntax error. I fixed that by using `in` precedence in all cases.

Next I decided that it's not so good to allow `for i = 1:n f(i) end`, so I deprecated that. Allowing space as an implicit separator here is one thing that prevents generalizing `in` syntax to other function names.

### CODE REVIEW

**Comment by tkelman** [comment]
> First, 2-argument comparisons are parsed as calls. Previously, we parsed `a<b` as `(comparison a < b)`, which I think is a quasi-bug since this can be handled as a simple function call and doesn't need chained comparison treatment.

Breaks-JuMP warning! And probably any other package that has macros that deal with comparisons.

**Comment by iamed2** [comment]
While working on https://github.com/invenia/JuliaFormat.jl I discovered the comparison special case and it surprised me. What was/is the reasoning/use for it?

**Comment by mlubin** [comment]
We can handle this breakage, don't let it hold you up.

**Comment by JeffBezanson** [comment]
Yes, everything in this PR is breaking. I think the new version is easier, since comparison chains `a < b < c` really are a different case than `a < b`. Any code that handles function calls can handle `a < b`, while comparison chains require inserting control flow, and might not be supported at all in a certain macro. So separating these cases more makes sense.

File: src/julia-parser.scm [hunk 1]
```diff
@@ -1399,14 +1400,20 @@
...
-           (error "invalid iteration specification")))))
+  (let* ((lhs (parse-pipes s))
+         (t   (peek-token s)))
+    (cond ((memq t '(= in ∈))
+           (take-token s)
+           (let* ((rhs (parse-pipes s))
+                  (t   (peek-token s)))
+             (if (not (or (closing-token? t) (newline? t)))
+                 ;; should be: (error "invalid iteration specification")
+                 (syntax-deprecation s (string "for " (deparse `(= ,lhs ,rhs)) " " t)
```
Comment (mlubin) [inline src/julia-parser.scm]: Any test for the deprecated syntax?

File: NEWS.md [hunk 2]
```diff
@@ -32,6 +32,15 @@ Language changes
   * A warning is always given when a method is overwritten (previously, this was done only when the new
     and old definitions were in separate modules) ([#14759]).
 
+  * `A <: B` is parsed as `Expr(:(<:), :A, :B)` in all cases ([#9503]). This also applies to the
+    `>:` operator.
+
+  * Simple 2-argument comparisons like `A < B` are parsed as calls intead of using the
+    `:comparison` expression type.
+
+  * A semicolon is now required when writing a `for` loop on one line, e.g.
```
Comment (innerlee) [inline NEWS.md]: The example has two semicolons.

**Comment by JeffBezanson** [comment]
I'm going to leave the for loop separator issue out of this; belongs in a separate change I think.

**Comment by tkelman** [comment]
@timholy is any change needed in `exprresolve_conditional` for this? https://github.com/JuliaLang/julia/blob/0d5f54132df14eb38847c65fcbdf4c3ac5011156/base/cartesian.jl#L371-L376

**Pull request merged**